src/v/datalake/coordinator/state_machine.cc

```diff
@@ -41,6 +44,45 @@ void maybe_log_update_error(
 coordinator_stm::coordinator_stm(ss::logger& logger, raft::consensus* raft)
   : raft::persisted_stm<>("datalake_coordinator_stm.snapshot", logger, raft) {}
 
+ss::future<checked<model::term_id, coordinator_stm::errc>>
+coordinator_stm::sync(model::timeout_clock::duration timeout) {
+    auto sync_res = co_await ss::coroutine::as_future(
+      raft::persisted_stm<>::sync(timeout));
+    if (sync_res.failed()) {
+        auto eptr = sync_res.get_exception();
+        auto msg = fmt::format("Exception caught while syncing: {}", eptr);
+        if (ssx::is_shutdown_exception(eptr)) {
+            vlog(datalake_log.debug, "Ignoring shutdown error: {}", msg);
+            co_return errc::shutting_down;
+        }
+        vlog(datalake_log.warn, "{}", msg);
+        co_return errc::replication_error;
+    }
+
+    if (!sync_res.get()) {
+        co_return errc::not_leader;
+    }
+    // At this point we're guaranteed that this node is the leader and that the
+    // STM has been caught up in the current term before (this doesn't mean the
+    // STM is currently caught up right now though!)
+    co_return _insync_term;
+}
+
+ss::future<checked<std::nullopt_t, coordinator_stm::errc>>
+coordinator_stm::replicate_and_wait(
+  model::term_id term, model::record_batch batch, ss::abort_source& as) {
+    auto opts = raft::replicate_options{raft::consistency_level::quorum_ack};
+    opts.set_force_flush();
+    auto res = co_await _raft->replicate(
+      term, model::make_memory_record_batch_reader(std::move(batch)), opts);
+    if (res.has_error()) {
+        co_return errc::replication_error;
+    }
+    auto committed_offset = _raft->committed_offset();
+    co_await wait_no_throw(committed_offset, ss::lowres_clock::now() + 30s, as);
```

Comment:
if there are multiple replicate requests pending it may wait for more than required, you can use `offset` from `replicate_result`

Reply:
Doen